opty/direct_collocation.py

```diff
@@ -458,23 +513,51 @@ def plot_constraint_violations(self, vector, axes=None):
         con_nodes = range(1, self.collocator.num_collocation_nodes)
 
         plot_inst_viols = self.collocator.instance_constraints is not None
+        num_inst_viols = len(instance_violations)
+
         if axes is None:
-            fig, axes = plt.subplots(1 + plot_inst_viols, squeeze=False,
-                                     layout='compressed')
+            fig, axes = plt.subplots(1 + num_plots, squeeze=False,
+                layout='compressed', figsize=(8, 2.0*(num_plots+1)))
+
         axes = axes.ravel()
 
         axes[0].plot(con_nodes, state_violations.T)
         axes[0].set_title('Constraint violations')
         axes[0].set_xlabel('Node Number')
         axes[0].set_ylabel('EoM violation')
 
+        # reduce the instance constrtaints to 2 significant digits.
+        instance_constr_plot = []
+        for exp1 in self.collocator.instance_constraints:
+            for a in sm.preorder_traversal(exp1):
+                if isinstance(a, sm.Float):
+                    exp1 = exp1.subs(a, round(a, 2))
+            instance_constr_plot.append(exp1)
```

Comment:
> Python has printing controls that work like this:
> 
> ```
> In [7]: '{:1.3f}'.format(1.2394359593939)
> Out[7]: '1.239'
> ```
> 
> is there a reason to not use something like that instead of literally rounding the numerical values?

Timo suggested this to me, so I used it, since ti worked. I had no idea how to do such things.

Reply:
Ok, I see now there may not be a straight forward way to use Python printing methods on SymPy Floats in an expression. I thought there should be a setting on the StrPrinter that would control display of decimals, but there doesn't seem to be one.

Reply:
Maybe Timo already showed this (I struggle to find his comment), but this is a way to control printing of floats:

```
In [47]: from mpmath.libmp import prec_to_dps, to_str as mlib_to_str

In [48]: class MyPrinter(sm.StrPrinter):
    ...:     _default_settings = {
    ...:         "order": None,
    ...:         "full_prec": "auto",
    ...:         "sympy_integers": False,
    ...:         "abbrev": False,
    ...:         "perm_cyclic": True,
    ...:         "min": None,
    ...:         "max": None,
    ...:         "dps": None,
    ...:     }
    ...: 
    ...:     def _print_Float(self, expr):
    ...:         prec = expr._prec
    ...:         if prec < 5:
    ...:             dps = 0
    ...:         else:
    ...:             dps = prec_to_dps(expr._prec)
    ...:         if self._settings['dps']:
    ...:             dps = self._settings['dps']
    ...:         if self._settings["full_prec"] is True:
    ...:             strip = False
    ...:         elif self._settings["full_prec"] is False:
    ...:             strip = True
    ...:         elif self._settings["full_prec"] == "auto":
    ...:             strip = self._print_level > 1
    ...:         low = self._settings["min"] if "min" in self._settings else None
    ...:         high = self._settings["max"] if "max" in self._settings else None
    ...:         rv = mlib_to_str(expr._mpf_, dps, strip_zeros=strip, min_fixed=low, max_fixed=high)
    ...:         if rv.startswith('-.0'):
    ...:             rv = '-0.' + rv[3:]
    ...:         elif rv.startswith('.0'):
    ...:             rv = '0.' + rv[2:]
    ...:         if rv.startswith('+'):
    ...:             # e.g., +inf -> inf
    ...:             rv = rv[1:]
    ...:         return rv
    ...: 

In [49]: MyPrinter(settings={'dps': 3}).doprint(f(1.329294))
Out[49]: 'f(1.33)'
```

Reply:
Hah, I found Timo's solution which was the same.

Reply:
> Maybe Timo already showed this (I struggle to find his comment), but this is a way to control printing of floats:
> 
> ```
> In [47]: from mpmath.libmp import prec_to_dps, to_str as mlib_to_str
> 
> In [48]: class MyPrinter(sm.StrPrinter):
>     ...:     _default_settings = {
>     ...:         "order": None,
>     ...:         "full_prec": "auto",
>     ...:         "sympy_integers": False,
>     ...:         "abbrev": False,
>     ...:         "perm_cyclic": True,
>     ...:         "min": None,
>     ...:         "max": None,
>     ...:         "dps": None,
>     ...:     }
>     ...: 
>     ...:     def _print_Float(self, expr):
>     ...:         prec = expr._prec
>     ...:         if prec < 5:
>     ...:             dps = 0
>     ...:         else:
>     ...:             dps = prec_to_dps(expr._prec)
>     ...:         if self._settings['dps']:
>     ...:             dps = self._settings['dps']
>     ...:         if self._settings["full_prec"] is True:
>     ...:             strip = False
>     ...:         elif self._settings["full_prec"] is False:
>     ...:             strip = True
>     ...:         elif self._settings["full_prec"] == "auto":
>     ...:             strip = self._print_level > 1
>     ...:         low = self._settings["min"] if "min" in self._settings else None
>     ...:         high = self._settings["max"] if "max" in self._settings else None
>     ...:         rv = mlib_to_str(expr._mpf_, dps, strip_zeros=strip, min_fixed=low, max_fixed=high)
>     ...:         if rv.startswith('-.0'):
>     ...:             rv = '-0.' + rv[3:]
>     ...:         elif rv.startswith('.0'):
>     ...:             rv = '0.' + rv[2:]
>     ...:         if rv.startswith('+'):
>     ...:             # e.g., +inf -> inf
>     ...:             rv = rv[1:]
>     ...:         return rv
>     ...: 
> 
> In [49]: MyPrinter(settings={'dps': 3}).doprint(f(1.329294))
> Out[49]: 'f(1.33)'
> ```

He did show this or something similar, but I liked the simple round(a, 3) much better! :-)